modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/ExecutionServiceImpl.java

```diff
@@ -0,0 +1,864 @@
+
+    /** */
+    private final QueryTaskExecutor taskExecutor;
+
+    /** */
+    private final AffinityService affSrvc;
+
+    /** */
+    private final MailboxRegistry mailboxRegistry;
+
+    /** */
+    private final MappingService mappingSrvc;
+
+    /** */
+    private final ExchangeService exchangeSrvc;
+
+    /** */
+    private final ClosableIteratorsHolder iteratorsHolder;
+
+    /** */
+    private final Map<UUID, QueryInfo> running;
+
+    /** */
+    private final RowHandler<Row> handler;
+
+    /** */
+    private final DdlSqlToCommandConverter ddlConverter;
+
+    public ExecutionServiceImpl(
+        TopologyService topSrvc,
+        MessageService msgSrvc,
+        QueryPlanCache planCache,
+        SchemaHolder schemaHolder,
+        QueryTaskExecutor taskExecutor,
+        RowHandler<Row> handler
+    ) {
+        this.handler = handler;
+        this.msgSrvc = msgSrvc;
+        this.schemaHolder = schemaHolder;
+        this.taskExecutor = taskExecutor;
+
+        locNodeId = topSrvc.localMember().id();
+        qryPlanCache = planCache;
+        running = new ConcurrentHashMap<>();
+        ddlConverter = new DdlSqlToCommandConverter();
+        iteratorsHolder = new ClosableIteratorsHolder(LOG);
+        mailboxRegistry = new MailboxRegistryImpl(topSrvc);
+        exchangeSrvc = new ExchangeServiceImpl(taskExecutor, mailboxRegistry, msgSrvc);
+        mappingSrvc = new MappingServiceImpl(topSrvc);
+        // TODO: fix this
+        affSrvc = cacheId -> Objects::hashCode;
+
+        topSrvc.addEventHandler(new NodeLeaveHandler(this::onNodeLeft));
+
+        init();
+    }
+
+    private void init() {
+        msgSrvc.register((n, m) -> onMessage(n, (QueryStartRequest) m), SqlQueryMessageGroup.QUERY_START_REQUEST);
+        msgSrvc.register((n, m) -> onMessage(n, (QueryStartResponse) m), SqlQueryMessageGroup.QUERY_START_RESPONSE);
+        msgSrvc.register((n, m) -> onMessage(n, (ErrorMessage) m), SqlQueryMessageGroup.ERROR_MESSAGE);
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<Cursor<List<?>>> executeQuery(
+        String schema,
+        String qry,
+        Object[] params
+    ) {
+        PlanningContext pctx = createContext(topologyVersion(), locNodeId, schema, qry, params);
+
+        List<QueryPlan> qryPlans = qryPlanCache.queryPlan(pctx, new CacheKey(pctx.schemaName(), pctx.query()), this::prepareQuery);
+
+        return executePlans(qryPlans, pctx);
+    }
+
+    /**
+     * Executes prepared plans.
+     * @param qryPlans Query plans.
+     * @param pctx Query context.
+     * @return List of query result cursors.
+     */
+    @NotNull public List<Cursor<List<?>>> executePlans(
+        Collection<QueryPlan> qryPlans,
+        PlanningContext pctx
+    ) {
+        List<Cursor<List<?>>> cursors = new ArrayList<>(qryPlans.size());
+
+        for (QueryPlan plan : qryPlans) {
+            UUID qryId = UUID.randomUUID();
+
+            Cursor<List<?>> cur = executePlan(qryId, pctx, plan);
```

Comment:
Looks like created cursors will remain open. Hope this place will be reworked under https://issues.apache.org/jira/browse/IGNITE-14749, otherwise I'll file a ticked for this